core/entity.js

```diff
@@ -1,285 +1,57 @@
-import { create_float_buffer, create_index_buffer } from './context';
-import { vec3, mat4, quat } from '../math';
-import { hex_to_rgb } from '../utils';
-
-// When using arrow keys for rotation simulate the mouse delta of this value.
-const KEY_ROTATION_DELTA = 3;
-
 const default_options = {
-  scale: vec3.unit.slice(),
-  position: vec3.zero.slice(),
-  rotation: quat.create(),
-
-  keyboard_controlled: false,
-  mouse_controlled: false,
-
-  move_speed: 3.5,
-  rotate_speed: .5,
-
-  dir_desc: {
-    87: 'f',
-    65: 'l',
-    68: 'r',
-    83: 'b',
-    69: 'u',
-    81: 'd',
-    38: 'pu',
-    40: 'pd',
-    39: 'yr',
-    37: 'yl'
-  },
-
+  components: [],
   skip: false,
 };
 
 export class Entity {
   constructor(options) {
-    this.matrix = mat4.create();
-    this.world_matrix = mat4.create();
-    this.world_to_self = mat4.create();
-
-    this._scale = vec3.unit.slice();
-    this._rotation = quat.create();
-
     Object.assign(this, default_options, options);
 
-    if (this.material) {
-      this.color_vec = [];
-      this.color_opacity = 1.0;
-      this.color = options.color || '#ffffff';
-    }
-
     this.entities = new Set();
+    const components = this.components;
+    this.components = new Map();
 
-    this.program = this.material && this.material.program;
-
-    if (this.vertices && this.vertices.length && this.indices && this.indices.length && this.normals && this.normals.length) {
-      this.create_buffers();
-    }
-  }
-
-  get up() {
-    const out = this.matrix.slice(4, 7);
-    return vec3.normalize(out, out);
-  }
-
-  get forward() {
-    const out = this.matrix.slice(8, 11);
-    return vec3.normalize(out, out);
-  }
-
-  set position(vec) {
-    mat4.compose(this.matrix, this.rotation, vec, this.scale);
+    this.add_components(components);
   }
 
-  get position() {
-    return this.matrix.slice(12, 15);
+  add_component(component) {
+    component.entity = this;
+    this.components.set(component.name, component);
   }
 
-  set scale(vec) {
-    this._scale = vec;
-    mat4.compose(this.matrix, this.rotation, this.position, vec);
+  add_components(components) {
```

Comment:
I don't like `add_components` and `get_components`. They both increase the API surface and the same can be achieved with their singular counterparts and a `for` loop on the user's side. I'm of the opinion that Cervus should only give the basic building blocks, or atoms, to developers, and encourage them to write helpers if they need them.

Reply:
so maybe just `get_components` ? this way it will return a single element array for one component.

Reply:
The singular version is more atomic and avoid the creation of the array if you only need on component (which should be most of the time?).